docs/architecture/adr-057-RPC.md

```diff
@@ -0,0 +1,97 @@
+# ADR 057: RPC
+
+## Changelog
+
+- 19-05-2020: created
+
+## Context
+
+Currently the RPC layer of Tendermint is using a variant of the JSON-RPC protocol. This ADR is meant to serve as a pro/con list for possible alternatives and JSON-RPC.
+
+There are currently two options being discussed: gRPC & JSON-RPC.
+
+### JSON-RPC:
+
+JSON-RPC is a JSON-based RPC protocol. Tendermint has implemented its own variant of JSON-RPC which is not compatible with the [JSON-RPC 2.0 specification](https://www.jsonrpc.org/specification).
+
+**Pros:**
+
+- Simple
+- Well-known and well-understood by users and integrators
+- Easy to interface with (HTTP and JSON)
+- Mature, common, and familiar tooling (e.g. `curl` and `jq`)
+- Integrates reasonably well with web infrastructure (proxies, API gateways, service meshes, caches, etc)
+
+**Cons:**
+
+- No schema support
+- RPC clients must be hand-written
+- Streaming not built into protocol
+- Underspecified types (e.g. numbers and timestamps)
+- Tendermint has its own implementation (not standards compliant, maintenance overhead)
+- Stdlib `jsonrpc` package only supports JSON-RPC 1.0, no dominant package for JSON-RPC 2.0
+- Tooling around documentation/specification (e.g. Swagger) could be better
+- JSON data is larger (offset by HTTP compression)
+- Serializing is slower ([~100% marshal, ~400% unmarshal](https://github.com/alecthomas/go_serialization_benchmarks)); insignificant in absolute terms
+
+### gRPC:
+
+gRPC is a high performant RPC framework. It has been battle tested by a large number of users and is heavily relied on and maintained by countless large corporations.
+
+**Pros:**
+
+- Efficient data retrieval for users, lite clients and other protocols
+- Easily implemented in supported languages (Go, Dart, JS, TS, rust, Elixir, Haskell, ...)
+- Can provide a REST API via [gRPC-gateway](https://github.com/grpc-ecosystem/grpc-gateway)
+- Can provide a GraphQL API via [Rejoiner](https://github.com/google/rejoiner)
+- Defined schema with richer type system (Protocol Buffers)
+- Can use common schemas and types across all protocols and data stores (RPC, ABCI, blocks, etc)
+- Established conventions for forwards- and backwards-compatibility
+- Bi-directional streaming
+- Servers and clients can be autogenerated in many languages (e.g. Tendermint-rs)
+
+**Cons:**
+
+- Users/integrators may not be familiar with gRPC and Protobuf; learning curve
+- Complex system involving cross-language schemas, code generation, and custom protocols
+- Tooling is less mature and harder to use (see e.g. `grpcurl`)
```

Comment:
Is this in comparison to `curl`? If you'll serve up a REST API with grpc-gateway, I don't see why this is a valid con

Reply:
In general, I'd say there's a lot more tooling for REST and JSON-RPC than gRPC. But you're right, if we offer both then it doesn't matter. The biggest tradeoff in that case would be maintenance overhead and complexity, but since we're relying on external packages it should be managable.

Reply:
I was trying to evaluate gRPC by itself, but it does make more sense to evaluate JSON-RPC vs gRPC + gRPC-gateway (REST). Ill amend the ADR to evaluate based off these two

>The biggest tradeoff in that case would be maintenance overhead and complexity,

what is the maintance overhead here? I feel like JSON-RPC has way more overhead since we have to maintain itself.

Reply:
Just having two API surfaces instead of one. We'd need documentation for both, testing for both, fix bugs and handle backwards compatibility in both, etc. But you're probably right that maintaining our own framework has more overhead.

Reply:
>Just having two API surfaces instead of one. We'd need documentation for both, testing for both, fix bugs and handle backwards compatibility in both, etc.

not sure if you have had a look at grpc-gateway but this is all automated. You add annotations to you service, like so https://github.com/tendermint/iavl/blob/bez%2F183-start-grpc-types/proto/iavl_api.proto#L15-L19, and the compiler will generate you swagger & the needed items for json-proxy. There isn't much manual interaction from the developer. this is essentially all that is needed: https://github.com/tendermint/iavl/blob/bez%2F183-start-grpc-types/cmd/iavlserver/main.go#L112-L141